gh-132661: Document t-strings and `templatelib` (#135229)
python/cpython: Doc/tutorial/inputoutput.rst

  but you'll also need to provide the information to be formatted. In the following code
  block there are two examples of how to format variables:

> 

Reviewer: Let's avoid unrelated changes

Author: Thanks; fixed. (A previous iteration of this PR _did_ make changes to this corner of `tutorial/inputoutput.rst`.)

   >>> repr((x, y, ('spam', 'eggs')))
   "(32.5, 40000, ('spam', 'eggs'))"

> The :mod:`string` module also contains support for so-called
> :ref:`$-strings <template-strings-pep292>` that offer yet another way to
> substitute values into strings, using placeholders like ``$x`` and replacing
> them with values from a dictionary. This syntax is easy to use, although
> it offers much less control of the formatting.

Reviewer 2: 'so called' to me reads a little dismissively. 

```suggestion
The :mod:`string` module contains support for a simple templating approach
based upon regular expressions, via :class:`string.Template`.
This offers yet another way to substitute values into strings,
using placeholders like ``$x`` and replacing them with values from a dictionary.
This syntax is easy to use, although it offers much less control for formatting.
```

on the ``=`` specifier. For a reference on these format specifications, see
the reference guide for the :ref:`formatspec`.

.. _tut-t-strings:

Template String Literals

> -------------------------

Reviewer 3: Could you please leave the tutorial to a separate PR (and replace `tut-t-strings` links with `t-strings` for now)?

This is a [Diátaxis](https://diataxis.fr/)-style tutorial;  IMO it would work better as an explanation and/or how-to guide.

Author: My instinct (and my read of review feedback so far) is that it feels like it fits well and follows nicely after the f-string section of the same doc (`tutorial/inputoutput.rst`). 

Leaving this open while we consider our options here.

Reviewer 3: I'll happily approve the rest of this PR.
For the tutorial I have more suggestions, so I'm asking to consider the options in a separate PR, to not block this one.

Author: Will do!

             for part in template
         )
     def user_html(user: User) -> Template:
         return t"<div><h1>{user.name}</h1></div>"

.. doctest::

     >>> evil_user = User(name="<script>alert('evil');</script>")
     >>> template = user_html(evil_user)
     >>> safe = sanitize_html_template(template)
     >>> print(safe)
     <div><h1>&lt;script&gt;alert(&#x27;evil&#x27;);&lt;/script&gt;</h1></div>

We are no longer vulnerable to XSS attacks because we are escaping the
interpolated values before they are included in the rendered HTML.

Of course, there's no need for code that processes :class:`!Template` instances
to be limited to returning a simple string. For instance, we could imagine
defining a more complex ``html()`` function that returns a structured
representation of the HTML:

     >>> from dataclasses import dataclass
     >>> from string.templatelib import Template
     >>> from html.parser import HTMLParser
     >>>
     >>> @dataclass
     ... class Element:
     ...     tag: str
     ...     attributes: dict[str, str]
     ...     children: list[str | Element]
     ...
     >>> def parse_html(template: Template) -> Element:
     ...     """

>      ...     Uses python's built-in HTMLParser to parse the template,

Reviewer 4: ```suggestion
     ...     Uses Python's built-in HTMLParser to parse the template,
```

     ...     handle any interpolated values, and return a tree of
     ...     Element instances.
     ...     """

>      ...     pass

Reviewer 4: I think the ellipsis better indicates "not yet implemented" than `pass` ("don't do anything, all is fine")

```suggestion
     ...     ...
```
